File: 141.Linked-List-Cycle/list.py
from typing import Optional


class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        if not head:
            return False

        slow = head
        fast = head.next

        # O(n)
        while slow and fast:
            if slow == fast:
                return True

            slow = slow.next
            temp = fast.next
            if not temp:
                return False

            fast = temp.next

        return False

File: 141.Linked-List-Cycle/list.c
#include <stdbool.h>

struct ListNode {
    struct ListNode *next;
};

bool hasCycle(struct ListNode *head) {
    if (!head) {
        return false;
    }

    struct ListNode *slow = head;
    struct ListNode *fast = head;

    // O(n)
    while (true) {
        if (!fast) return false;
        fast = fast->next;
        
        if (!fast) return false;
        fast = fast->next;

        slow = slow->next;
        if (slow == fast) {
            return true;
        }
    }

    return false;
}
